collector/netdev_common.go

```diff
@@ -106,5 +110,68 @@ func (c *netDevCollector) Update(ch chan<- prometheus.Metric) error {
 			ch <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, float64(value), dev)
 		}
 	}
+	if *netdevAddressInfo {
+		interfaces, err := net.Interfaces()
+		if err != nil {
+			return fmt.Errorf("could not get network interfaces: %w", err)
+		}
+
+		desc := prometheus.NewDesc(prometheus.BuildFQName(namespace, "network_address",
+			"info"), "node network address by device",
+			[]string{"device", "address", "netmask", "scope"}, nil)
+
+		for _, addr := range getAddrsInfo(interfaces) {
+			ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1,
+				addr.device, addr.addr, addr.netmask, addr.scope)
+		}
+	}
 	return nil
 }
+
+type addrInfo struct {
+	device  string
+	addr    string
+	scope   string
+	netmask string
+}
+
+func scope(ip net.IP) string {
+	if ip.IsLoopback() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() {
+		return "link-local"
+	}
+
+	if ip.IsInterfaceLocalMulticast() {
+		return "interface-local"
+	}
+
+	if ip.IsGlobalUnicast() {
+		return "global"
+	}
+
+	return ""
+}
+
+// getAddrsInfo returns interface name, address, scope and netmask for all interfaces.
+func getAddrsInfo(interfaces []net.Interface) []addrInfo {
+	var res []addrInfo
```

Comment:
Rather than use append, this can be a little more efficient if you do something like this:

```go
res := make([]addrInfo, len(interfaces))
```

Then when you do the for loop

```go
for i, interface := range interfaces {
...
  res[i] = addrInfo{
...
```

When you use append like this, it's causing a lot of memory alloc/free churn.

Reply:
Unfortunately the size of res slice can't be taken from `len(interfaces)` because interface can return multiple addresses.

```go
func (ifi *Interface) Addrs() ([]Addr, error) {...}
```

So we need to traverse all interfaces to know the size of res slice. But in the same time we want to collect all addresses information and put it to res slice. And it will be inconvenient to traverse them twice, probably we don't win anything here. But I think we can preallocate more memory for slice capacity as a tradeoff. I think it should elimintate some allocations, maybe it will be good enough?

```go
res := make([]addrInfo, 0, 16)
```

Reply:
Alright, I guess it's fine as-is.